File: 3499-maximize-active-section-with-trade-i/3499-maximize-active-section-with-trade-i.go
func maxActiveSectionsAfterTrade(s string) int {
	count1 := 0
	for i := 0; i < len(s); i++ {
		if s[i] == '1' {
			count1++
		}
	}

	zeroBlocks := []int{}
	i := 0
	for i < len(s) {
		start := i
		for i < len(s) && s[i] == s[start] {
			i++
		}
		if s[start] == '0' {
			zeroBlocks = append(zeroBlocks, i-start)
		}
	}

	if len(zeroBlocks) < 2 {
		return count1
	}

	best := 0
	for j := 1; j < len(zeroBlocks); j++ {
		best = max(best, zeroBlocks[j]+zeroBlocks[j-1])
	}

	return count1 + best
}

// greedy
// time: O(n)
// space: O(n)

File: 3499-maximize-active-section-with-trade-i/3499-maximize-active-section-with-trade-i.py
class Solution:
    def maxActiveSectionsAfterTrade(self, s: str) -> int:
        count1 = 0
        for i in range(len(s)):
            if s[i] == '1':
                count1 += 1
        
        zero_blocks = []
        idx = 0
        while idx < len(s):
            start = idx
            while idx < len(s) and s[idx] == s[start]:
                idx += 1
            if s[start] == '0':
                zero_blocks.append(idx - start)
        
        if len(zero_blocks) < 2:
            return count1
        
        best = 0
        for i in range(1, len(zero_blocks), 1):
            best = max(best, zero_blocks[i] + zero_blocks[i-1])
        
        return best + count1

# greedy
# time: O(n)
# space: O(n)